### PR TITLE
update resources faster with package_update

### LINKED ISSUE #5713: Very high render times on resource addition and updates
**CKAN version**

CKAN 2.8.4

**Describe the bug**

Updating resources, or adding resources, for a dataset that already has a large number of resources, is unmanageably slow.

**Steps to reproduce**

- Create a dataset with the resources from https://www.data.qld.gov.au/dataset/ambient-estuarine-water-quality-monitoring-data-near-real-time-sites-2012-to-present-day
- Edit and save a resource

**Expected behavior**

Resources should update within a few seconds.

**Additional details**

We recently installed the ckanext-archiver, ckanext-report, and ckanext-qa extensions; it is possible that there is an interaction with these extensions that is causing or exacerbating the problem.

Debugging suggests that the slowness is occurring in `ckan/logic/action/update.py`, within `package_update`, when generating a new revision. This may be indicative of a database deadlock.

### PR DESCRIPTION
**Pull request by wardi**
Fixes #5713

### Proposed fixes:

1. selectively skip validation and dictization on resources if both are true:
   - dataset metadata is unchanged
   - skipped resource metadata is unchanged
2. skip updates completely if no metadata was changed
3. compare new and existing resource model objects before commit to avoid commit processing time
4. pass `original_package` in context from actions that call `package_show` with `for_update=True` to `package_update` to allow low-cost validated metadata comparisons

| 100-resource dataset action | old | new | improvement |
| --- | --- | --- | --- |
| resource_create | .721 | .419 | 70% |
| resource_update | 1.05 | .445 | 130% |
| resource_delete | .709 | .478 | 48% |
| package_resource_reorder | .697 | .479 | 45% |
| package_patch | .660 | .621 | 6% * |
| resource_patch | .746 | .413 | 80% |
| package_revise (updating resource) | .787 | .267 | 190% |

\* validation and dictization applied for all resources when any dataset field changes, we only save commit time

> [!NOTE] 
> This change will affect custom validation rules that access resource metadata from dataset or other resource validators:
> - only the changed resources are passed to validation
> - flattened data the validators receive won't include skipped resources
>
> This change does *not* affect validation of resource fields that depend on package fields because the package fields are always available in flattened data

### Features:

- [x] includes tests covering changes
- [ ] includes updated documentation
- [ ] includes user-visible changes
- [ ] includes API changes
- [ ] includes bugfix for possible backport

### CODE REVIEW

File: ckan/logic/action/update.py [hunk 1]
```diff
@@ -296,16 +326,54 @@ def package_update(
...
     log.debug('package_update validate_errs=%r user=%s package=%s data=%r',
               errors, user, context['package'].name, data)
 
     if errors:
         model.Session.rollback()
         raise ValidationError(errors)
 
-    #avoid revisioning by updating directly
+    # merge skipped resources back into validated data
+    if copy_resources:
```
Comment (amercader) [inline ckan/logic/action/update.py:359]: Can you explain this snippet? I get that we want to merge the skipped resources back with the changed ones but I don't understand the second while loop

**Review by amercader — COMMENTED** [review]
I think I finally understand this (mostly, see comment :slightly_smiling_face:  ) 
The numbers are clear and these changes are worth merging but we should be mindful that we are adding more and more logic to `package_update` so need to make sure is maintainable going forward.

File: ckan/logic/__init__.py [hunk 2]
```diff
@@ -555,6 +555,7 @@ def get_action(action: str) -> Action:
     # wrap the functions
     for action_name, _action in _actions.items():
         def make_wrapped(_action: Action, action_name: str):
+            @functools.wraps(_action)
```
[the remark this answers is not in the document]
Comment (wardi) [reply ckan/logic/__init__.py:558]: Yes, just trying to do things the standard way if we can

File: ckan/logic/action/create.py [hunk 3]
```diff
@@ -293,44 +291,23 @@ def resource_create(context: Context,
     for plugin in plugins.PluginImplementations(plugins.IResourceController):
         plugin.before_resource_create(context, data_dict)
 
-    if 'resources' not in pkg_dict:
-        pkg_dict['resources'] = []
-
-    upload = uploader.get_resource_uploader(data_dict)
```
[the remark this answers is not in the document]
Comment (wardi) [reply ckan/logic/action/create.py:299]: Yes I assume the current code isn't uploading twice but didn't check.

File: ckan/logic/action/update.py (hunk 1 above)
Comment (wardi) [reply ckan/logic/action/update.py:359]: the second while loop adds anything remaining in copy_resources after the end of `data['resources']` is reached

File: ckan/lib/dictization/model_save.py [hunk 4]
```diff
@@ -52,16 +52,17 @@ def resource_dict_save(res_dict: dict[str, Any],
...
 
-    session.add(obj)
     return obj
 
 
 def package_resource_list_save(
         res_dicts: Optional[list[dict[str, Any]]],
-        package: 'model.Package', context: Context) -> None:
+        package: 'model.Package', context: Context,
+        copy_resources: dict[int, int] | tuple[()]) -> None:
```
[the remark this answers is not in the document]
Comment (wardi) [reply ckan/lib/dictization/model_save.py:65]: copy_resources is passed an empty tuple as a default parameter here:
https://github.com/ckan/ckan/pull/8360/files#diff-65184b94c9686447997dab4cfd41411d58f1d83033a870345baad20e11bdbd33R276

 Python doesn't have a frozendict and I don't want to use a mutable value as a default.

**Pull request merged**